### ISSUE TITLE
Routing Priority doesn't seem to work as expected

### ISSUE THREAD

**Issue by sychan** [containous/traefik#6433]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

I have a "default" handler for a path prefix which spawns a new container that has a longer routing rule that matches a cookie. The cookie is sent back to the browser and the browser reloads the page. The expected behavior is that the new container with a longer routing rule will match for any request that presents the cookie, and the shorter rule will only match for requests without a matching cookie.

Here is the routing rule for the "default" container:
~~~
traefik.http.routers.authsvc.rule=Host("oursite.org.tld") && PathPrefix("/narrative")
~~~

Here is the routing rule for the container that should handle the a request with a cookie with the value 497ff413a8f4b3782f44efb5c614a6bc:
~~~
traefik.http.routers.sychan.rule=Host("oursite.org.tld") && PathPrefix("/narrative") && HeadersRegexp("Cookie",`497ff413a8f4b3782f44efb5c614a6bc`)
~~~

I've also tried explicitly assigning a lower priority value to the default container labels with settings like this:
~~~
traefik.http.routers.authsvc.priority=1
~~~

And then assigning a much higher value to the cookie matching container withL
~~~
traefik.http.routers.sychan.priority=5000
~~~

### What did you expect to see?

I expected that:
1. The longer rule would match with higher priority for any request that matches the rules
2. When an explicit priority is assigned to the containers via labels, the higher priority container with a rule match would be selected.

### What did you see instead?

The default container with the shorter rule kept matching. I have verified that traefik saw both the default and the cookie matching container, and that the cookie was being presented by the browser.

When I deleted the default container, then traefik would *finally* route to the longer rule.

If I deleted the old containers and create new ones with the added .priority labels, traefik would still route to the shorter matching rule.

This only started occurring after I upgraded from 2.1.2 to 2.1.6

### Output of `traefik version`: (_What version of Traefik are you using?_)

~~~
/ # traefik version
Version:      2.1.6
Codename:     cantal
Go version:   go1.13.8
Built:        2020-02-28T17:40:18Z
OS/Arch:      linux/amd64
/ #
~~~

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

These are being created in Rancher via the GUI (and for the cookie matching container, via the Rancher API), here is the dump of the equivalent docker-compose from Rancher (I have stripped out sensitive configuration vars, or host identifying info)

```yaml
version: '2'
services:
  Traefik:
    image: traefik:v2.1
    hostname: artifactory
    stdin_open: true
    tty: true
    ports:
    - 80:80/tcp
    - 8080:8080/tcp
    command:
    - --accesslog=true
    - --log.level=DEBUG
    - --api.insecure=true
    - --providers.rancher=true
    - --providers.rancher.exposedByDefault=false
    - --entrypoints.web.address=:80
    - --metrics=true
    - --metrics.prometheus=true
    labels:
      io.rancher.container.pull_image: always
  authsvc:
    image: kbase/narrative-traefiker:latest
    stdin_open: true
    tty: true
    ports:
    - 80:5000/tcp
    labels:
      traefik.enable: 'true'
      traefik.http.routers.authsvc.rule: Host("host.domain.tld") && PathPrefix("/narrative")
      traefik.http.routers.authsvc.entrypoints: web
      io.rancher.container.pull_image: always
  narrative-sychan:
    image: kbase/narrative:latest
    stdin_open: true
    tty: true
    ports:
    - 8888:8888/tcp
    labels:
      session_id: 497ff413a8f4b3782f44efb5c614a6bc
      traefik.enable: 'True'
      traefik.http.routers.sychan.rule: Host("host.domain.tld") && PathPrefix("/narrative") && HeadersRegexp("Cookie",`497ff413a8f4b3782f44efb5c614a6bc`)
      traefik.http.routers.sychan.entrypoints: web
      io.rancher.container.pull_image: always

```

### If applicable, please paste the log output in DEBUG level (`--log.level=DEBUG` switch)

```
(paste your output here)
```

**Comment by jbdoumenjou** [comment]
Hi @sychan ,

I tried to reproduce with this configuration:
```
version: '3.5'

services:

  traefik:
    image: traefik:v2.1.6
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    command:
      - --entrypoints.web.address=:80
      - --log.level=DEBUG
      - --api.insecure=true
      - --providers.docker
      - --providers.docker.exposedByDefault=false
    ports:
      - '80:80'
      - '8080:8080'
    networks:
      net:
        aliases:
          - traefik

  api1:
    image: containous/whoami
    labels:
      traefik.enable: 'true'
      traefik.http.routers.router0.entrypoints: web
      traefik.http.routers.router0.rule: Host(`whoami`)
#      traefik.http.routers.router0.priority: 41
    ports:
      - '8001:80'
    networks:
      net:
        ipv4_address: 10.0.1.3

  api2:
    image: containous/whoami
    labels:
      traefik.enable: 'True'
      traefik.http.routers.router1.entrypoints: web
      traefik.http.routers.router1.rule: Host(`whoami`) && HeadersRegexp("Cookie",`298zf09hf012fh2`)
#      traefik.http.routers.router1.priority: 420
    ports:
      - '8002:80'
    networks:
      net:
        ipv4_address: 10.0.1.4

  api3:
    image: containous/whoami
    labels:
      traefik.enable: 'True'
      traefik.http.routers.router2.entrypoints: web
      traefik.http.routers.router2.rule: Host(`whoami`) && HeadersRegexp("Cookie",`298zf09hf012fh3`)
#      traefik.http.routers.router2.priority: 420
    ports:
      - '8003:80'
    networks:
      net:
        ipv4_address: 10.0.1.5

networks:
  net:
    name: "net"
    driver: bridge
    ipam:
      config:
        - subnet: 10.0.1.0/28

```
I use the IP address of the whoami services to see which router is used and I changed the priority values to check if they are taken into account.

I used curl:
```curl whoami -H "Cookie: foo=298zf09hf012fh2; bar=u32t4o3tb3gg43;"```

And so on for each valid header value and the right router is responding.
So, I didn't find any issue.

As you used the HeadersRegExp, any matching value will be routed, if the request doesn't have a Cookie header containing the value, the "default" router will manage the request as expected.
Could you double check your configuration and your Traefik logs ?

**Comment by sychan** [comment]
I have verified my configuration, and the problem remained, I verified that there was only 1 cookie with a matching hash string. Here are the request headers passed (the [DELETED] value is the auth token, and does not match the randomly generated narrative_session cookie)
~~~
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
Accept-Encoding: gzip, deflate
Accept-Language: en-US,en;q=0.9
Cache-Control: no-cache
Connection: keep-alive
Cookie: _ga=GA1.2.1110700235.1567008955; _xsrf=2|ec2365a6|4c89aac7b1f6763864b26cf2fdcf4a49|1582742074; _gcl_au=1.1.829714086.1582826335; narrative_session=497ff413a8f4b3782f44efb5c614a6bc; kbase_session=[DELETED]
Host: host.domain.tld
Pragma: no-cache
Referer: http://host.domain.tld/narrative/ws.12202.obj.1
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.122 Safari/537.36
~~~

   In any case, the problem is that the HeaderRegExp() clause seems to be being ignored, and the shortest rule was always the match. So another cookie having a matching value is really the opposite type of problem.

  However I tried reverting to v2.1.2 and changed the configuration to write the logs to files so that it would be easier to share, and after a restart the routing works. The only differences are the traefik version and the logging directives. So the problem is intermittent, or else tied to v2.1.6 - here is the current config that works:
~~~
version: '2'
services:
  Traefik:
    image: traefik:v2.1.2
    hostname: artifactory
    stdin_open: true
    tty: true
    ports:
    - 80:80/tcp
    - 8080:8080/tcp
    command:
    - --log.level=DEBUG
    - --log.filePath=/var/log/traefik.log
    - --accesslog=true
    - --accesslog.filepath=/var/log/access.log
    - --api.insecure=true
    - --providers.rancher=true
    - --providers.rancher.exposedByDefault=false
    - --entrypoints.web.address=:80
    - --metrics=true
    - --metrics.prometheus=true
    labels:
      io.rancher.container.pull_image: always
  authsvc:
    image: kbase/narrative-traefiker:latest
    environment:
    stdin_open: true
    tty: true
    ports:
    - 80:5000/tcp
    labels:
      traefik.enable: 'true'
      traefik.http.routers.authsvc.rule: Host("host.domain.tld") && PathPrefix("/narrative")
      traefik.http.routers.authsvc.entrypoints: web
      io.rancher.container.pull_image: always
  narrative-sychan:
    image: kbase/narrative:latest
    stdin_open: true
    tty: true
    ports:
    - 8888:8888/tcp
    labels:
      session_id: 497ff413a8f4b3782f44efb5c614a6bc
      traefik.enable: 'True'
      traefik.http.routers.sychan.rule: Host("host.domain.tld") && PathPrefix("/narrative") && HeadersRegexp("Cookie",`497ff413a8f4b3782f44efb5c614a6bc`)
      traefik.http.routers.sychan.entrypoints: web
      io.rancher.container.pull_image: always

~~~

    Can you try your test by adding in a PathPrefix clause as well and see if that effects things? I will keep trying with v2.1.6 to see if the problem returns.

**Comment by rtribotte** [comment]
Hi @sychan,

I just tested with PathPrefix.
I used this config:
```
version: '3.5'

services:

  traefik:
    image: traefik:v2.1.6
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    command:
      - --entrypoints.web.address=:80
      - --log.level=DEBUG
      - --api.insecure=true
      - --providers.docker
      - --providers.docker.exposedByDefault=false
    ports:
      - '80:80'
      - '8080:8080'
    networks:
      net:
        aliases:
          - traefik

  api1:
    image: containous/whoami
    labels:
      traefik.enable: 'true'
      traefik.http.routers.router0.entrypoints: web
      traefik.http.routers.router0.rule: Host(`whoami`)
    ports:
      - '8001:80'
    networks:
      net:
        ipv4_address: 10.0.1.3

  api2:
    image: containous/whoami
    labels:
      traefik.enable: 'True'
      traefik.http.routers.router1.entrypoints: web
      traefik.http.routers.router1.rule: Host(`whoami`) && PathPrefix("/narrative")
    ports:
      - '8002:80'
    networks:
      net:
        ipv4_address: 10.0.1.4

  api3:
    image: containous/whoami
    labels:
      traefik.enable: 'True'
      traefik.http.routers.router2.entrypoints: web
      traefik.http.routers.router2.rule: Host(`whoami`) && PathPrefix("/narrative") && HeadersRegexp("Cookie",`497ff413a8f4b3782f44efb5c614a6bc`)
    ports:
      - '8003:80'
    networks:
      net:
        ipv4_address: 10.0.1.5

networks:
  net:
    name: "net"
    driver: bridge
    ipam:
      config:
        - subnet: 10.0.1.0/28
```
With curl:
```
curl whoami/narrative -H "Cookie: foo=298zf09hf012fh2; narrative_session=497ff413a8f4b3782f44efb5c614a6bc; bar=u32t4o3tb3gg43;"
```

This behave as expected, the longest rule applies.

Could you provide your traefik logs ?

**Comment by sychan** [comment]
I think I have figured out the problem, and it isn't traefik (its user error, grrr).

The default container and traefik were both binding to port 80 in the managed rancher network, and it looks like it was a race condition to see which server ended up servicing incoming requests. I realized this when I looked more carefully at the traefik access logs and saw that it was being bypassed on the requests that were broken.

Sorry for wasting your time, and thank you are being patient with me as I worked this out.
